src/utils/UnitSpherical/point.jl

```diff
@@ -0,0 +1,163 @@
+        UnitSphericalPoint{T}(v[1], v[2], v[3])
+    elseif length(v) == 2
+        UnitSphereFromGeographic()(v)
+    else
+        @boundscheck begin
+            throw(ArgumentError("""
+            Passed a vector of length `$(length(v))` to the `UnitSphericalPoint` constructor, 
+            which only accepts vectors of lengths: 
+            - **3** (assumed to be on the unit sphere) 
+            - **2** (assumed to be geographic lat/long)
+            """))
+        end
+    end
+end
+
+Base.show(io::IO, p::UnitSphericalPoint) = print(io, "UnitSphericalPoint($(p.x), $(p.y), $(p.z))")
+
+# ## Interface implementations
+
+# StaticArraysCore.jl interface implementation
+Base.setindex(p::UnitSphericalPoint, args...) = throw(ArgumentError("`setindex!` on a UnitSphericalPoint is not permitted as it is static."))
+StaticArrays.similar_type(::Type{<: UnitSphericalPoint}, ::Type{Eltype}, ::Size{(3,)}) where Eltype = UnitSphericalPoint{Eltype}
+# Base math implementation (really just forcing re-wrapping)
+# Base.:(*)(a::UnitSphericalPoint, b::UnitSphericalPoint) = a .* b
+function Base.broadcasted(f, a::AbstractArray{T}, b::UnitSphericalPoint) where {T <: UnitSphericalPoint}
+    return Base.broadcasted(f, a, (b,))
+end
+Base.isnan(p::UnitSphericalPoint) = any(isnan, p)
+Base.isinf(p::UnitSphericalPoint) = any(isinf, p)
+Base.isfinite(p::UnitSphericalPoint) = all(isfinite, p)
+
+
+# GeoInterface implementation
+## Traits:
+GI.trait(::UnitSphericalPoint) = GI.PointTrait()
+GI.geomtrait(::UnitSphericalPoint) = GI.PointTrait()
+## Coordinate traits:
+GI.is3d(::GI.PointTrait, ::UnitSphericalPoint) = true
+GI.ismeasured(::GI.PointTrait, ::UnitSphericalPoint) = false
+## Accessors:
+GI.ncoord(::GI.PointTrait, ::UnitSphericalPoint) = 3
+GI.getcoord(::GI.PointTrait, p::UnitSphericalPoint) = p[i]
+## Metadata (CRS, extent, etc)
+GI.crs(::UnitSphericalPoint) = GFT.ProjString("+proj=cart +R=1 +type=crs") # TODO: make this a full WKT definition
+# TODO: extent is a little tricky - do we do a spherical cap or an Extents.Extent?
+
+# ## Spherical distance
+"""
+    spherical_distance(x::UnitSphericalPoint, y::UnitSphericalPoint)
+
+Compute the spherical distance between two points on the unit sphere.  
+Returns a `Number`, usually Float64 but that depends on the input type.
+
+# Extended help
+
+## Doctests
+
+```jldoctest
+julia> spherical_distance(UnitSphericalPoint(1, 0, 0), UnitSphericalPoint(0, 1, 0))
+1.5707963267948966
+```
+
+```jldoctest
+julia> spherical_distance(UnitSphericalPoint(1, 0, 0), UnitSphericalPoint(1, 0, 0))
+0.0
+```
+"""
+spherical_distance(x::UnitSphericalPoint, y::UnitSphericalPoint) = acos(clamp(x ⋅ y, -1.0, 1.0))
+
+# ## Random points
+Random.rand(rng::Random.AbstractRNG, ::Random.SamplerType{UnitSphericalPoint}) = rand(rng, UnitSphericalPoint{Float64})
+function Random.rand(rng::Random.AbstractRNG, ::Random.SamplerType{UnitSphericalPoint{T}}) where T <: Number
+    θ = 2π * rand(rng, T)
+    ϕ = acos(2 * rand(rng, T) - 1)
```

Comment:
I was a bit confused here since I think the convention is that `phi` resembles the longitudes and `theta` latitudes, so maybe the variable names should be exchanged here for better readability.